netx/conn.go

```diff
@@ -0,0 +1,150 @@
+
+func (r *roundRobinWriter) Write(byte []byte) (int, error) {
+
+	if r.instances == nil {
+		r.updateInstances(<-r.provider)
+	}
+
+	select {
+	case newInstances := <-r.provider:
+		r.updateInstances(newInstances)
+		return r.write(byte)
+	default:
+		return r.write(byte)
+	}
+}
+
+func (r *roundRobinWriter) updateInstances(newInstances []net.Addr) {
+	r.instances = make(chan net.Addr, len(newInstances))
+	for _, instance := range newInstances {
+		r.instances <- instance
+	}
+}
+
+func (r *roundRobinWriter) write(payload []byte) (int, error) {
+	// Read next instance from queue
+	instance := <-r.instances
+	// Enqueue instance for round robin behaviour
+	r.instances <- instance
+
+	return r.sender(instance, payload)
+}
+
+// NetworkSend is a Sender that can write payload to the network address
+func NetworkSend(addr net.Addr, payload []byte) (int, error) {
+	conn, err := net.Dial(addr.Network(), addr.String())
+	if err != nil {
+		return 0, err
+	}
+	defer func() {
+		err := conn.Close()
+		if err != nil {
+			logrus.WithError(err).Warn("There was an error when closing connection")
+		}
+	}()
+
+	n, err := conn.Write(payload)
+	if err != nil {
+		return 0, fmt.Errorf("could not sent payload to %s://%s: %s", addr.Network(), addr.String(), err)
+	}
+	return n, nil
+}
+
+// DiscoveryServiceInstanceProvider returns InstanceProvider that is updated with list of instances in interval
+func DiscoveryServiceInstanceProvider(serviceName string, interval time.Duration, client DiscoveryServiceClient) InstanceProvider {
+	instancesChan := make(chan []net.Addr, 1)
+
+	go func() {
+		for range time.NewTicker(interval).C {
+			instances, err := client.GetAddrsByName(serviceName)
+			if err != nil {
+				continue
+			}
+			//TODO(janisz): Update instances only when they changed.
+			instancesChan <- instances
+		}
+	}()
+
+	return instancesChan
+}
+
+// DiscoveryServiceClient represents generic discovery service client that can return list of services
+type DiscoveryServiceClient interface {
+	// GetAddrsByName returns list of services with given name
+	GetAddrsByName(serviceName string) ([]net.Addr, error)
+}
+
+// NewConsulDiscoveryServiceClient returns DiscoverServiceClient backed by Consul
+func NewConsulDiscoveryServiceClient(client *api.Client) DiscoveryServiceClient {
+	return &consulDiscoveryServiceClient{
+		client: client,
+	}
+}
+
+type consulDiscoveryServiceClient struct {
+	client *api.Client
+}
+
+func (c *consulDiscoveryServiceClient) GetAddrsByName(serviceName string) ([]net.Addr, error) {
+	//TODO(janisz): Add fallback to other datacenters with query
+	services, _, err := c.client.Catalog().Service(serviceName, "", nil)
+
+	if err != nil {
+		return nil, fmt.Errorf("could NOT find service in Consul: %s", err)
+	}
+
+	instances := make([]net.Addr, len(services))
+	for i, instance := range services {
+		//TODO(janisz): take network from configuration
+		instances[i] = tcpAddr(fmt.Sprintf("%s:%d", instance.ServiceAddress, instance.ServicePort))
```

Comment:
Make it configurable now (i.e. as parameter in structure creating function) or change default to UDP - especially because we need a UDP not TCP here (hardcoded value is wrong for us).